sklearn/metrics/cluster/_supervised.py

```diff
@@ -1275,12 +1275,14 @@ def entropy(labels):
     -----
     The logarithm used is the natural logarithm (base-e).
     """
-    xp, is_array_api_compliant = get_namespace(labels)
+    xp, is_array_api_compliant, device_ = get_namespace_and_device(labels)
     labels_len = labels.shape[0] if is_array_api_compliant else len(labels)
     if labels_len == 0:
         return 1.0
 
-    pi = xp.astype(xp.unique_counts(labels)[1], xp.float64)
```

Comment:
Before deciding to conditional upcast based on the maximum precision float dtype supported by the namespace & device, I think we need to do some "git blame" archeology to understand why if this upcast was first introduced and if it is still actually needed or not.

One want to find out empirically would be to craft some synthetic input arrays to trigger numerical problems with `float32` operations only.

If it is needed, maybe it's possible do some trick similar to the log sum exp numerical stability trick:

https://gregorygundersen.com/blog/2020/02/09/log-sum-exp/

I don't have a pen and paper / whiteboard at hand today to check myself though.

Reply:
that's fun! It looks like an activity for the weekend. Let me know if you (or anyone that is reading this) discover something before that

Reply:
After some digging, I found that:
- the calculation of `pi` has kind of always been casted to float [source](https://github.com/scikit-learn/scikit-learn/commit/c3d3906a918d661c54918708135e2448c2fcbc87#diff-8841e9c8a04424f47364ab93b8f9c4b238eee3ad5c9b25e21424488b27bf28cdR741-R742). There is a [commit of a day prior](https://github.com/scikit-learn/scikit-learn/commit/2eb983342bfda3520aef3babfaddcfed7f0609f0#diff-8841e9c8a04424f47364ab93b8f9c4b238eee3ad5c9b25e21424488b27bf28cdR748) where it was not, but it was quickly changed. I couldn't understand why tbh, do you know why it could be?
- when `np.bincount` was introduced, it was cast to `np.float` [source](https://github.com/scikit-learn/scikit-learn/commit/8cf5896729ea903293c52ce1acb4f5545ef1eee0#diff-8841e9c8a04424f47364ab93b8f9c4b238eee3ad5c9b25e21424488b27bf28cdR782)
- the alias was changed to `np.float64` [here](https://github.com/scikit-learn/scikit-learn/commit/268b6ea160d373d86b26f7292a90a5b800c8c213#diff-8841e9c8a04424f47364ab93b8f9c4b238eee3ad5c9b25e21424488b27bf28cdR867)

Reply:
regarding

> One way to find out empirically would be to craft some synthetic input arrays to trigger numerical problems with float32 operations only.

I may need some pointers 😕 I will check the article later this week

Reply:
Thanks for the investigation. Indeed the choice of `np.float32` vs `np.float64` was not done to avoid numerical stability problems with `np.float32` but just that we need a floating point conversion at some point (from the integer input data) and that `np.float64` is the default floating point dtype in numpy.

So I think it's ok to use the `_max_precision_float_dtype` for this.